tests/ops/op_math/test_controlled_op.py

```diff
@@ -0,0 +1,688 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pytest
+from copy import copy
+from scipy import sparse
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.operation import DecompositionUndefinedError, GeneratorUndefinedError
+from pennylane.ops.op_math.controlled_class import Controlled, ControlledOperation
+from pennylane.wires import Wires
+
+
+from gate_data import CNOT, CZ, CSWAP, Toffoli, CRotx, CRoty, CRotz, CRot3
+
+base_num_control_mats = [
+    (qml.PauliX("a"), 1, CNOT),
+    (qml.PauliZ("a"), 1, CZ),
+    (qml.SWAP(("a", "b")), 1, CSWAP),
+    (qml.PauliX("a"), 2, Toffoli),
+    (qml.RX(1.234, "b"), 1, CRotx(1.234)),
+    (qml.RY(-0.432, "a"), 1, CRoty(-0.432)),
+    (qml.RZ(6.78, "a"), 1, CRotz(6.78)),
+    (qml.Rot(1.234, -0.432, 9.0, "a"), 1, CRot3(1.234, -0.432, 9.0)),
+]
+
+
+class TempOperator(qml.operation.Operator):
```

Comment:
I think it would be cleaner just to import `Operator` and `Operation`.

Reply:
only really gets used in one place...